ccl/sqlproxyccl: TestProxyProtocol failed

ccl/sqlproxyccl.TestProxyProtocol [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20642294?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20642294?buildTab=artifacts#/) on release-25.4 @ [b4669b97748b6f9d2d9a0ec1a7368c5fb9464f4e](https://github.com/cockroachdb/cockroach/commits/b4669b97748b6f9d2d9a0ec1a7368c5fb9464f4e):


```
=== RUN   TestProxyProtocol
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/2dcdeaae7bc06474bb2d1d585750c46b/logTestProxyProtocol3778243654
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    proxy_handler_test.go:235: B
    proxy_handler_test.go:254: 
        	Error Trace:	pkg/ccl/sqlproxyccl/proxy_handler_test.go:2945
        	Error:      	Expected nil, but got: &sqlproxyccl.errWithCode{
        	            	    code:  4,
        	            	    cause: &withstack.withStack{
        	            	        cause: &errutil.withPrefix{
        	            	            cause:  &errors.errorString{s:"invalid length of startup packet: 369295617"},
        	            	            prefix: "while receiving startup message",
        	            	        },
        	            	        stack: &withstack.stack{0x4884651, 0x4890cac, 0x4896860, 0x18cd218, 0x493994},
        	            	    },
        	            	}
        	Test:       	TestProxyProtocol
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/2dcdeaae7bc06474bb2d1d585750c46b/logTestProxyProtocol3778243654
--- FAIL: TestProxyProtocol (2.34s)
=== RUN   TestProxyProtocol/server_requires_proxy_protocol
    testing.go:1577: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestProxyProtocol/server_requires_proxy_protocol (0.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #133186 ccl/sqlproxyccl: TestProxyProtocol failed [C-test-failure O-robot T-serverless branch-release-24.3]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestProxyProtocol.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55739

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -162,7 +162,7 @@ func TestProxyProtocol(t *testing.T) {
 	sqlDB.Exec(t, `CREATE USER bob WITH PASSWORD 'builder'`)
 
 	var validateFn func(h *proxyproto.Header) error
-	withProxyProtocol := func(p bool) (server *Server, addrs *serverAddresses) {
+	withProxyProtocol := func(p bool, stopper *stop.Stopper) (server *Server, addrs *serverAddresses) {
 		options := &ProxyOptions{
 			RoutingRule:          ts.AdvSQLAddr(),
 			SkipVerify:           true,
@@ -171,7 +171,7 @@ func TestProxyProtocol(t *testing.T) {
 		options.testingKnobs.validateProxyHeader = func(h *proxyproto.Header) error {
 			return validateFn(h)
 		}
-		return newSecureProxyServer(ctx, t, sql.Stopper(), options)
+		return newSecureProxyServer(ctx, t, stopper, options)
 	}
 
 	timeout := 3 * time.Second
@@ -228,11 +228,10 @@ func TestProxyProtocol(t *testing.T) {
 	})()
 
 	testSQLNoRequiredProxyProtocol := func(s *Server, addr string) {
-		url := fmt.Sprintf("postgres://bob:builder@%s/tenant-cluster-42.defaultdb?sslmode=require", addr)
+		url := fmt.Sprintf("postgres://bob:builder@%s/tenant-cluster-42.defaultdb?sslmode=require&sslrootcert=%s", addr, datapathutils.TestDataPath(t, "testserver.crt"))
 		// No proxy protocol.
 		te.TestConnect(ctx, t, url,
 			func(conn *pgx.Conn) {
-				t.Log("B")
 				require.NotZero(t, s.metrics.CurConnCount.Value())
 				require.NoError(t, runTestQuery(ctx, conn))
 			},
@@ -247,7 +246,7 @@ func TestProxyProtocol(t *testing.T) {
 	}
 
 	testSQLRequiredProxyProtocol := func(s *Server, addr string) {
-		url := fmt.Sprintf("postgres://bob:builder@%s/tenant-cluster-42.defaultdb?sslmode=require", addr)
+		url := fmt.Sprintf("postgres://bob:builder@%s/tenant-cluster-42.defaultdb?sslmode=require&sslrootcert=%s", addr, datapathutils.TestDataPath(t, "testserver.crt"))
 		// No proxy protocol.
 		_ = te.TestConnectErr(ctx, t, url, codeClientReadFailed, "tls error")
 		// Proxy protocol.
@@ -264,7 +263,10 @@ func TestProxyProtocol(t *testing.T) {
 	}
 
 	t.Run("server doesn't require proxy protocol", func(t *testing.T) {
-		s, addrs := withProxyProtocol(false)
+		ctx := context.Background()
+		stopper := stop.NewStopper()
+		defer stopper.Stop(ctx)
+		s, addrs := withProxyProtocol(false, stopper)
 		// Test SQL on the default listener. Both should go through.
 		testSQLNoRequiredProxyProtocol(s, addrs.listenAddr)
 		// Test SQL on the proxy protocol listener. Only request with PROXY should go
@@ -279,7 +281,10 @@ func TestProxyProtocol(t *testing.T) {
 	})
 
 	t.Run("server requires proxy protocol", func(t *testing.T) {
-		s, addrs := withProxyProtocol(true)
+		ctx := context.Background()
+		stopper := stop.NewStopper()
+		defer stopper.Stop(ctx)
+		s, addrs := withProxyProtocol(true, stopper)
 		// Test SQL on the default listener. Both should go through.
 		testSQLRequiredProxyProtocol(s, addrs.listenAddr)
 		// Test SQL on the proxy protocol listener. Only request with PROXY should go

